punchcut/src/source/glyf/scaler.rs

```diff
@@ -0,0 +1,621 @@
+                            (p1.x.wrapping_sub(p2.x), p1.y.wrapping_sub(p2.y))
+                        }
+                    };
+                    if dx != 0 || dy != 0 {
+                        for p in &mut outline.points[start_point..end_point] {
+                            p.x += dx;
+                            p.y += dy;
+                        }
+                    }
+                }
+                if self.hint {
+                    let ins = composite.instructions().unwrap_or_default();
+                    // TODO: variations
+                    // self.deltas.resize(delta_base, Point::new(0, 0));
+                    if !ins.is_empty() {
+                        // Append the current phantom points to the outline.
+                        self.push_phantom(outline);
+                        // For composite glyphs, the unscaled and original points are simply
+                        // copies of the current point set.
+                        self.scaler.context.unscaled.clear();
+                        self.scaler
+                            .context
+                            .unscaled
+                            .extend_from_slice(&outline.points[point_base..]);
+                        self.scaler.context.original.clear();
+                        self.scaler
+                            .context
+                            .original
+                            .extend_from_slice(&outline.points[point_base..]);
+                        let point_end = outline.points.len();
+                        // Round the phantom points.
+                        for p in &mut outline.points[point_end - 4..] {
+                            p.x = math::round(p.x);
+                            p.y = math::round(p.y);
+                        }
+                        // Clear the "touched" flags that are used during IUP processing.
+                        const TOUCHED_FLAGS: u8 = 0x08 | 0x10;
+                        for tag in &mut outline.tags[point_base..] {
+                            *tag &= !TOUCHED_FLAGS;
+                        }
+                        if !self.hint(outline, point_base, contour_base, ins, true) {
+                            return Err(Error::HintingFailed(glyph_id));
+                        }
+                        // As in simple outlines, drop the phantom points.
+                        outline.points.truncate(outline.points.len() - 4);
+                        outline.tags.truncate(outline.tags.len() - 4);
+                    }
+                }
+            }
+        }
+        if recurse_depth == 0 {
+            // If the first phantom point is non-zero, shift the entire outline.
+            let pp0x = self.phantom[0].x;
+            if pp0x != 0 {
+                for p in &mut outline.points {
+                    p.x -= pp0x;
+                }
+            }
+            outline.is_scaled = self.scaler.is_scaled;
+        }
+        Ok(())
+    }
```

Comment:
it would  be nice if we could figure out a way to split this function up a bit, although I appreciate all the comments.

Reply:
I can definitely split this up into functions for simple and composite glyphs and that should make each more tractable.